sycl/include/CL/sycl/detail/device_impl.hpp

```diff
@@ -21,44 +20,141 @@ namespace sycl {
+  /// For host device an exception is thrown
+  ///
+  /// @return non-constant reference to PI device
+  RT::PiDevice &getHandleRef() {
+    if (MIsHostDevice)
+      throw invalid_object_error("This instance of device is a host instance");
+
+    return MDevice;
+  }
+
+  /// Get constant reference to PI device
+  ///
+  /// For host device an exception is thrown
+  ///
+  /// @return constant reference to PI device
+  const RT::PiDevice &getHandleRef() const {
+    if (MIsHostDevice)
+      throw invalid_object_error("This instance of device is a host instance");
+
+    return MDevice;
+  }
+
+  /// Check if device is a host device
+  ///
+  /// @return true if SYCL device is a host device
+  bool is_host() const { return MIsHostDevice; }
+
+  /// Check if device is a CPU device
+  ///
+  /// @return true if SYCL device is a CPU device
+  bool is_cpu() const { return (MType == PI_DEVICE_TYPE_CPU); }
+
+  /// Check if device is a GPU device
+  ///
+  /// @return true if SYCL device is a GPU device
+  bool is_gpu() const { return (MType == PI_DEVICE_TYPE_GPU); }
+
+  /// Check if device is an accelerator device
+  ///
+  /// @return true if SYCL device is an accelerator device
+  bool is_accelerator() const { return (MType == PI_DEVICE_TYPE_ACC); }
+
+  /// Get associated SYCL platform
+  ///
+  /// If this SYCL device is an OpenCL device then the SYCL platform
+  /// must encapsulate the OpenCL cl_plaform_id associated with the
+  /// underlying OpenCL cl_device_id of this SYCL device. If this SYCL device
+  /// is a host device then the SYCL platform must be a host platform.
+  /// The value returned must be equal to that returned
+  /// by get_info<info::device::platform>().
+  ///
+  /// @return The associated SYCL platform.
+  platform get_platform() const;
+
+  /// Check SYCL extension support by device
+  ///
+  /// @param ExtensionName is a name of queried extension.
+  /// @return true if SYCL device supports the extension.
+  bool has_extension(const string_class &ExtensionName) const;
 
-  virtual vector_class<device> create_sub_devices(size_t nbSubDev) const = 0;
-
-  virtual vector_class<device>
-  create_sub_devices(const vector_class<size_t> &counts) const = 0;
-
-  virtual vector_class<device>
-  create_sub_devices(info::partition_affinity_domain affinityDomain) const = 0;
-
-  static vector_class<device>
-  get_devices(info::device_type deviceType = info::device_type::all);
+  vector_class<device>
+  create_sub_devices(const cl_device_partition_property *Properties,
+                     size_t SubDevicesCount) const;
 
+  /// Partition device into sub devices
+  ///
+  /// If this SYCL device does not support info::partition_property::partition_equally
+  /// a feature_not_supported exception must be thrown.
+  ///
+  /// @param ComputeUnits is a desired count of compute units in each sub device.
+  /// @return A vector class of sub devices partitioned from this SYCL device
+  /// equally based on the ComputeUnits parameter.
```

Comment:
TBH, I can't figure out the meaning of "equally based". Could you please explain or reformulate?